components/network/openvpn/patches/02-ipsec-bypass.patch

```diff
@@ -12,22 +12,32 @@
 + * IPsec.)
 + */
 +static void
-+set_ipsec_bypass(int sock)
++set_ipsec_bypass(int sock, const sa_family_t ai_family)
 +{
 +  ipsec_req_t ipsr;
++  int result = 0;
 +
 +  /* Don't bother if the socket is a failure.  Caller never checks... */
 +  if (sock == -1)
 +    return;
 +
++  /* Don't bother if it is not IPv4/IPv6 socket... */
++  if (ai_family != AF_INET && ai_family != AF_INET6)
++    return;
++
 +  ipsr.ipsr_ah_req = IPSEC_PREF_NEVER;
 +  ipsr.ipsr_esp_req = IPSEC_PREF_NEVER;
 +  ipsr.ipsr_self_encap_req = IPSEC_PREF_NEVER;
 +  ipsr.ipsr_auth_alg = 0;
 +  ipsr.ipsr_esp_alg = 0;
 +  ipsr.ipsr_esp_auth_alg = 0;
 +
-+  if (setsockopt(sock, IPPROTO_IP, IP_SEC_OPT, &ipsr, sizeof (ipsr)) == -1) {
++  if (ai_family == AF_INET) {
++    result = setsockopt(sock, IPPROTO_IP, IP_SEC_OPT, &ipsr, sizeof (ipsr));
++  } else { /* AF_INET6 */
++    result = setsockopt(sock, IPPROTO_IPV6, IPV6_SEC_OPT, &ipsr, sizeof (ipsr));
++  }
++  if (result == -1) {
 +    switch (errno) {
 +    case EPROTONOSUPPORT:
 +      dmsg (D_SOCKET_DEBUG, "Not using IPsec.");
```

Comment:
"IPsec not loaded" is a more accurate description.

Reply:
Have one more question... If I receive EPROTONOSUPPORT error in zone, but it works in GZ.... Does this mean that zone misses some privileges?

Reply:
Also if it fails in zone with EPROTONOSUPPORT, in GZ when run as usual user, it will fail with EPERM (so tests will fail). The question is if we should also handle this fail gracefully (I mean, ignore EPERM as we do with EPROTONOSUPPORT)?

Reply:
See here:

http://src.illumos.org/source/xref/illumos-gate/usr/src/uts/common/inet/ip/ip.c#6158

It's possible your GZ has indicated IPsec is loaded (because ANY ONE of your zones used IPsec at some point) and your NGZ has not.  Each IPsec netstack has ipsec_loader_state in it.

Kernel calls to ipsec_loader_loadnow() toggle this state:

http://src.illumos.org/source/s?refs=ipsec_loader_loadnow&project=illumos-gate

This is why I asked about running "ipseckey flush" earlier.

Reply:
yes, after ipseckey flush it fails differently - ENOPERM.

Reply:
So clearly we do the is-loaded check BEFORE the privilege check.  A regular user will need the IP_CONFIG privilege to bypass.  (I thought openvpn needed this privilege anyway?!)

Reply:
well, at least openvpn tests run without root privileges (without this patch). It can do simple client-server pings...